### ISSUE TITLE
Reduction getting stuck on large ranges

### ISSUE THREAD

**Issue by ikabadzhov** [AdaptiveCpp/AdaptiveCpp#1496]
**Bug summary**

I just pulled latest master, and attempted to run a code including SYCL (sum) reduction. I noticed that the computation was taking unexpectedly long (and I never waited for it to finish). Instead, I created a small repro and noticed that after some threshold size, even an empty kernel "carrying" a reduction objection would take forever to execute.

**To Reproduce**

A minimal repro:

```cpp
#include <sycl/sycl.hpp>

int main(int argc, char **argv) {

  sycl::queue queue;

  unsigned long long res = 0;
  sycl::buffer<unsigned long long> res_b(&res, 1);

  long long len = (argc > 1) ? atoll(argv[1]) : 1 << 24;

  queue.submit([&](sycl::handler &cgh) {
    auto sum_reduction = sycl::reduction(res_b, cgh, sycl::plus<>());
    cgh.parallel_for(sycl::range<1>(len), sum_reduction,
                     [=](sycl::id<1> item, auto &sum) {});
  });
  queue.wait();

  return 0;
}
```

I didn't pass any extra flags to the ACPP build. Then, I created my executable as:
acpp --acpp-targets="generic" -O3  reduce.cpp -o red_acpp

And simply run:
for p in {1..5}; do time ./red_acpp $((1<<(14 + $p))); done

The first 3 runs are fast as expected, but then I am stuck, so Ctrl+C:
```
real    0m0,397s
user    0m0,246s
sys     0m0,151s

real    0m0,348s
user    0m0,220s
sys     0m0,135s

real    0m0,318s
user    0m0,229s
sys     0m0,092s
^C
real    12m59,571s
user    0m0,422s
sys     0m0,124s
```

**Additional diagnostic information**

Few trivial observations:

* I noticed that ACPP recognizes my CPU both as OpenCL as well as OpenMP backend. Same things happens on both backends;
* Substituting sycl::plus with std::plus is of no help;
* Switching to nd_range is of no help;
* Buffer pointing to an array of 1 element instead of a reference is of no help.
* It suffices to get rid of the reduction objects in the cgh.parallel_for to get the expected behaviour, i.e. this doesn't get stuck:
```
  queue.submit([&](sycl::handler &cgh) {
    auto sum_reduction = sycl::reduction(res_b, cgh, sycl::plus<>());
    cgh.parallel_for(sycl::range<1>(len), [=](sycl::id<1> item) {}); });
```

**Comment by illuhad** [comment]
Can you try using USM instead of buffers? There are additional complexities with buffers.

USM is in almost any case the superior memory management model anyway, as buffers are known to add potentially substiantial overheads independently of the SYCL implementation.

**Comment by ikabadzhov** [comment]
would that mean substituting `sycl::buffer<unsigned long long> res_b(&res, 1);`, with a malloc shared? but then the reduction `sycl::reduction(res_b, cgh, sycl::plus<>())` expects a buffer, and indeed I get plenty of errors of not matched accessor constructor, if I substitute the first argument in the reduction with a shared array. I am speculating, I did not understand you - could you please elaborate.

**Comment by illuhad** [comment]
> would that mean substituting sycl::buffer<unsigned long long> res_b(&res, 1);, with a malloc shared? but then the reduction sycl::reduction(res_b, cgh, sycl::plus<>()) expects a buffer, and indeed I get plenty of errors of not matched accessor constructor, if I substitute the first argument in the reduction with a shared array. I am speculating, I did not understand you - could you please elaborate.

I would strongly recommend learning about the SYCL USM memory model. It's the way to get performance with any SYCL application.

No, USM does not necessarily mean `malloc_shared`. Explicit USM device allocations (`malloc_device`) as well as host-resident allocations (`malloc_host`) are also a part of USM. The terminology is a bit confusing: In SYCL, USM just means "pointer-based memory management". `malloc_shared`, where allocations automatically migrate, is just one form of USM. `malloc_device` is similar to `cudaMalloc`, if you are familiar with CUDA.

Reductions have an overload for USM pointers. USM does not need the `handler`, so you will need to remove that in your call.
This is what this application looks like with USM device allocations - and indeed, with USM it works, so indeed the problem is related to complexities around the buffer model as I suspected:

```c++

#include <sycl/sycl.hpp>

int main(int argc, char **argv) {

  sycl::queue q;

  unsigned long long res = 0;
  unsigned long long* res_buff = sycl::malloc_device<unsigned long long>(1, q);

  q.memcpy(res_buff, &res, sizeof(unsigned long long)).wait();
  
  long long len = (argc > 1) ? atoll(argv[1]) : 1 << 24;

  q.parallel_for(len, sycl::reduction(res_buff, sycl::plus<>()),
                 [=](auto id, auto& sum){});
  q.wait();

  return 0;
}
```

**Comment by illuhad** [comment]
Potentially related to #1476

**Comment by ikabadzhov** [comment]
Thanks a lot for the detailed answer @illuhad. Before closing this issue, I would like to point out that when using the latest release 24.02.0, **both the buffer/accessor as well as** the USM version snippet would fail to compile as:
```
error: too few template arguments for class template 'plus'
  q.parallel_for(len, sycl::reduction(res_buff, sycl::plus<>()),
                                                      ^
[]/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/libkernel/functional.hpp:37:30: note: template is declared here
template <typename T> struct plus {
~~~~~~~~~~~~~~~~~~~~~        ^
1 error generated.
```

Okay, so I tried `sycl::plus<unsigned long long>` which would fail to compile with a very long error message starting with:
```
In file included from reduce.cpp:1:
In file included from []/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/sycl.hpp:31:
In file included from []/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/sycl.hpp:93:
In file included from[]/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/queue.hpp:51:
In file included from []/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/handler.hpp:67:
In file included from []/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/hipSYCL/glue/kernel_launcher_factory.hpp:57:
[]/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/hipSYCL/glue/llvm-sscp/sscp_kernel_launcher.hpp:163:7: error: no matching function for call to object of type 'const (lambda at red_acpp.cpp:15:18)'
      _k(this_item);
      ^~
```

On the other hand, in the develop branch, both `sycl::plus<>` as well as `sycl::plus<unsigned long long>` work as expected.

**Comment by illuhad** [comment]
Reduction support with `generic` target was only added after the 24.02 release in #1453. So this is expected.
